discovery/dns/dns_test.go

```diff
@@ -178,3 +179,96 @@ func TestDNS(t *testing.T) {
 		})
 	}
 }
+
+func TestSDConfigUnmarshalYAML(t *testing.T) {
+
+	marshal := func(c SDConfig) []byte {
+		d, err := yaml.Marshal(c)
+		if err != nil {
+			panic(err)
+		}
+		return d
+	}
+
+	unmarshal := func(d []byte) func(interface{}) error {
+		return func(o interface{}) error {
+			return yaml.Unmarshal(d, o)
+		}
+	}
+
+	cases := []struct {
+		name      string
+		input     SDConfig
+		expectErr bool
+	}{
+		{
+			name: "valid srv",
+			input: SDConfig{
+				Names: []string{"a.example.com", "b.example.com"},
+				Type:  "SRV",
+			},
+			expectErr: false,
+		},
+		{
+			name: "valid a",
+			input: SDConfig{
+				Names: []string{"a.example.com", "b.example.com"},
+				Type:  "A",
+				Port:  5300,
+			},
+			expectErr: false,
+		},
+		{
+			name: "valid aaaa",
+			input: SDConfig{
+				Names: []string{"a.example.com", "b.example.com"},
+				Type:  "AAAA",
+				Port:  5300,
+			},
+			expectErr: false,
+		},
+		{
+			name: "invalid a without port",
+			input: SDConfig{
+				Names: []string{"a.example.com", "b.example.com"},
+				Type:  "A",
+			},
+			expectErr: true,
+		},
+		{
+			name: "invalid aaaa without port",
+			input: SDConfig{
+				Names: []string{"a.example.com", "b.example.com"},
+				Type:  "AAAA",
+			},
+			expectErr: true,
+		},
+		{
+			name: "invalid empty names",
+			input: SDConfig{
+				Names: []string{},
+				Type:  "AAAA",
+			},
+			expectErr: true,
+		},
+		{
+			name: "invalid unkown dns type",
+			input: SDConfig{
+				Names: []string{"a.example.com", "b.example.com"},
+				Type:  "PTR",
+			},
+			expectErr: true,
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			var config SDConfig
+			d := marshal(c.input)
+			err := config.UnmarshalYAML(unmarshal(d))
+			if (err != nil) != c.expectErr {
+				t.Errorf("test case %q: UnmarshalYAML expect error: %v, got: %v; %v", c.name, c.expectErr, err != nil, err)
```

Comment:
no need to repeat `c.name` in the error message since it is already passed to `t.Run`.

Reply:
ok.